### PR TITLE
Frontend/feature/image upload

### PR DESCRIPTION
**Pull request by MichaelGuldborg**
I've added functionality for updating the user profile picture in the UserForm

I've added a component that handles taking an image file input (ImageInput)
And added a component for showing a circular avatar image (AvatarInput)
I've added a service function to upload images to the backend

**Waiting** for #704 to be merged

### CODE REVIEW

File: app/frontend/src/components/AvatarInput.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import React from "react";
+
+import {ImageInput, ImageInputProps} from "./ImageInput";
+
+export interface AvatarInputProps extends ImageInputProps {
+    src?: string;
+    username?: string;
+}
+
+const AvatarInput: React.FC<AvatarInputProps> = ({id, name, username, src, ...rest}) => {
```
Comment (lucaslcode) [inline app/frontend/src/components/AvatarInput.tsx]:
```suggestion
function AvatarInput({id, name, username, src, ...rest}: AvatarImageProps) {
```
style nit

File: app/frontend/src/components/AvatarInput.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    return (
+        <ImageInput id={id} name={name} {...rest}>
+            <IconButton component="span">
+                <Avatar
+                    src={src}
+                    imgProps={{style: {objectFit: 'contain'}}}
+                    style={{height: 120, width: 120, objectFit: 'contain'}}
```
Comment (lucaslcode) [inline app/frontend/src/components/AvatarInput.tsx]: We tend to use jss (`const useStyle = makeStyles((theme) => ({...}))`) rather than inline styles (unless there is a reason you wanted inline here?)

Also I think it would be good if the parent could set the size (perhaps by this component accepting a `className` prop) rather than it being fixed.

File: app/frontend/src/components/ImageInput.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import React from "react";
+
+export interface ImageInputProps {
+    id?: string;
+    name?: string;
+    onChange: (file: File) => void;
+}
+
+export const ImageInput: React.FC<ImageInputProps> = ({id, name, children, onChange}) => {
```
Comment (lucaslcode) [inline app/frontend/src/components/ImageInput.tsx]:
```suggestion
export default function ImageInput({id, name, children, onChange}: ImageInputProps) {
```

File: app/frontend/src/components/ImageInput.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    const inputId = id + '-file-input';
+    return (
+        <React.Fragment>
```
Comment (lucaslcode) [inline app/frontend/src/components/ImageInput.tsx]:
```suggestion
        <>
```
shorthand

File: app/frontend/src/components/ImageInput.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+                        return;
+                    }
+                    const file = event.target.files[0];
+                    onChange && onChange(file);
+                }}
+            />
+            <label htmlFor={inputId}>
+                {children}
+            </label>
+        </React.Fragment>
```
Comment (lucaslcode) [inline app/frontend/src/components/ImageInput.tsx]:
```suggestion
        </>
```

File: app/frontend/src/components/ImageInput.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    onChange: (file: File) => void;
+}
+
+export const ImageInput: React.FC<ImageInputProps> = ({id, name, children, onChange}) => {
+    const inputId = id + '-file-input';
+    return (
+        <React.Fragment>
+            <input
+                accept="image/*"
+                style={{display: 'none'}}
```
Comment (lucaslcode) [inline app/frontend/src/components/ImageInput.tsx]: use `className`
(Hmm if this is the only style in the file, maybe you don't have to).

File: app/frontend/src/components/ImageInput.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                    onChange && onChange(file);
+                }}
+            />
+            <label htmlFor={inputId}>
+                {children}
+            </label>
+        </React.Fragment>
+    )
+}
+export default ImageInput;
```
Comment (lucaslcode) [inline app/frontend/src/components/ImageInput.tsx]: Newline. There are some other formatting things too, run `yarn format`

File: app/frontend/src/features/profile/ProfileForm.tsx [hunk 8]
```diff
@@ -87,12 +93,34 @@ export default function EditProfileForm() {
...
+                      id={'profile-picture'}
+                      name={'image'}
+                      src={value}
+                      onChange={async (file) => {
+                          // TODO move media endpoint to config
+                        const response = await service.api.uploadFile(file);
+                        const thumbnailUrl = 'https://dev-user-media.coucher.org/media/img/avatar/' + response.key + '.jpg';
+                        setValue('avatarUrl',thumbnailUrl)
+                      }}
+                  />
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/ProfileForm.tsx]: I think you could move this all into the AvatarInput component. Not sure if others agree or not.

That way you don't have to specify onChange and there would be one place to change the upload logic.

Comment (lucaslcode) [reply app/frontend/src/features/profile/ProfileForm.tsx]: Additionally, if you use a [react-query mutation](https://react-query.tanstack.com/guides/mutations) you'll get free loading state

**Review by lucaslcode — DISMISSED** [review]
Nice one

Some small style nitpicks and an optional idea to move the logic into the component. IT would also be great to have the new components in storybook!

**Comment by lucaslcode** [comment]
> Aapeli agrees that it would be better that the url is part of the response from the backend rather than the way I've done it currently

What do you mean by this?

**Comment by lucaslcode** [comment]
Also the upload is not saved to the profile currently, it needs to be added to the UpdateProfileReq request in `service/user.ts`

File: app/frontend/src/components/ImageInput.tsx (hunk 6 above)
Comment (andreaowu) [reply app/frontend/src/components/ImageInput.tsx]: In order to make the code organized, we should just keep it consistent and just use `className`.

File: app/frontend/src/features/profile/ProfileForm.tsx [hunk 9]
```diff
@@ -77,6 +80,9 @@ export default function EditProfileForm() {
     updateUserProfile({ profileData: data, setMutationError: setErrorMessage });
   });
 
+
+
+
```
Comment (andreaowu) [inline app/frontend/src/features/profile/ProfileForm.tsx]: nit: delete extra lines

File: app/frontend/src/features/profile/ProfileForm.tsx [hunk 10]
```diff
@@ -87,12 +93,34 @@ export default function EditProfileForm() {
       {user ? (
         <>
           <form onSubmit={onSubmit}>
+
+
```
Comment (andreaowu) [inline app/frontend/src/features/profile/ProfileForm.tsx]: nit: delete extra lines

File: app/frontend/src/features/profile/ProfileForm.tsx [hunk 11]
```diff
@@ -87,12 +93,34 @@ export default function EditProfileForm() {
...
+            <Controller
+                name={'avatarUrl'}
+                control={control}
+                render={({value}) => {
+                  return <AvatarInput
+                      id={'profile-picture'}
+                      name={'image'}
+                      src={value}
+                      onChange={async (file) => {
+                          // TODO move media endpoint to config
```
Comment (andreaowu) [inline app/frontend/src/features/profile/ProfileForm.tsx]: nit: indentation

File: app/frontend/src/features/profile/ProfileForm.tsx [hunk 12]
```diff
@@ -87,12 +93,34 @@ export default function EditProfileForm() {
...
+                        const response = await service.api.uploadFile(file);
+                        const thumbnailUrl = 'https://dev-user-media.coucher.org/media/img/avatar/' + response.key + '.jpg';
+                        setValue('avatarUrl',thumbnailUrl)
+                      }}
+                  />
+                }}
+            />
+
+
```
Comment (andreaowu) [inline app/frontend/src/features/profile/ProfileForm.tsx]: nit: delete extra lines

File: app/frontend/src/features/profile/ProfileForm.tsx (hunk 8 above)
Comment (andreaowu) [reply app/frontend/src/features/profile/ProfileForm.tsx]: +1 otherwise every time we have to upload, this code needs to be copied again (basically what Lucas said)

File: app/frontend/src/components/AvatarInput.tsx (hunk 2 above)
Comment (MichaelGuldborg) [reply app/frontend/src/components/AvatarInput.tsx]: I think that's a good idea, I'll change it

File: app/frontend/src/components/ImageInput.tsx (hunk 6 above)
Comment (MichaelGuldborg) [reply app/frontend/src/components/ImageInput.tsx]: I'll change it and remember to use className in the future aswell

File: app/frontend/src/features/profile/ProfileForm.tsx (hunk 8 above)
Comment (MichaelGuldborg) [reply app/frontend/src/features/profile/ProfileForm.tsx]: I totally agree, I'll move this logic into the ImageInput component.
I'm not sure how to avoid having the onChange since I need to use the setValue function to set the thumbnail url on the form, right?

**Comment by MichaelGuldborg** [comment]
> Also the upload is not saved to the profile currently, it needs to be added to the UpdateProfileReq request in service/user.ts

I've added the avatarKey and avatarUrl to the request values but it still does not seem to work

File: app/frontend/src/features/profile/ProfileForm.tsx [hunk 13]
```diff
@@ -87,12 +95,32 @@ export default function EditProfileForm() {
...
+              render={({ value }) => {
+                return (
+                  <AvatarInput
+                    className={classes.avatar}
+                    id={"profile-picture"}
+                    name={"image"}
+                    src={value}
+                    onChange={async (values) => {
+                      setValue("avatarUrl", values.thumbnail_url);
+                      setValue("avatarKey", values.key);
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/ProfileForm.tsx]: Ok after thinking about it, I think the way to do this is:
- only have `avatarKey` here in the form
- handle the `avatarUrl` and displaying it within the `AvatarInput` with local state
- the [`Controller`](https://react-hook-form.com/api/#Controller) render function also has `onChange(newValue)` as a parameter, so pass that to `AvatarInput`
- then within `AvatarInput` in the "real" `onChange`, you can set the url in the local component state, then call from the props `onChange(avatarKey)`

Let me know if that makes no sense :P

File: app/frontend/src/service/user.ts [hunk 14]
```diff
@@ -49,7 +49,8 @@ export type UpdateUserProfileData = Pick<
   | "countriesVisited"
   | "countriesLived"
   | "additionalInformation"
->;
+  | "avatarKey"
```
Comment (lucaslcode) [inline app/frontend/src/service/user.ts:51]: You need to apply the new data to the gRPC request message in this file in `updateProfile` function

File: app/frontend/src/features/profile/ProfileForm.tsx (hunk 13 above)
Comment (MichaelGuldborg) [reply app/frontend/src/features/profile/ProfileForm.tsx]: It makes sense,
I've fixed it now 👌

**Comment by MichaelGuldborg** [comment]
@aapeliv 
The updateUser request gives an error when it includes the avatarKey
And some of the uploaded images gives "Internal Server Error" when I request them after upload

File: app/frontend/src/components/AvatarInput.tsx [hunk 15]
```diff
@@ -24,16 +26,23 @@ export function AvatarInput({
   src,
   username,
   className,
+  onChange,
   ...rest
 }: AvatarInputProps) {
   const classes = useStyles();
+  const [imageUrl, setImageUrl] = useState(src);
+  const handleChange = (values: ImageInputValues) => {
+    const randomInt = Math.floor(Math.random() * 100); // force reload onChange
```
Comment (lucaslcode) [inline app/frontend/src/components/AvatarInput.tsx]: Did you have problems when not doing this? I would have thought every upload has a different key and therefore a different url

File: app/frontend/src/components/AvatarInput.tsx [hunk 16]
```diff
@@ -24,16 +26,23 @@ export function AvatarInput({
...
+    setImageUrl(values.thumbnail_url + "?rand=" + randomInt);
+    onChange && onChange(values.key);
```
Comment (lucaslcode) [inline app/frontend/src/components/AvatarInput.tsx]:
```suggestion
    onChange?.(values.key);
```
You don't have to change this but I just learnt this recently and I think it's cool :D

File: app/frontend/src/features/profile/ProfileForm.tsx [hunk 17]
```diff
@@ -67,7 +68,7 @@ export default function EditProfileForm() {
       lat: user?.lat ?? undefined,
       lng: user?.lng ?? undefined,
       radius: user?.radius ?? undefined,
-      // avatarUrl: user?.avatarUrl ?? undefined,
+      avatarKey: undefined,
```
Comment (lucaslcode) [inline app/frontend/src/features/profile/ProfileForm.tsx]: Does this need to be `avatarKey: user.avatarKey` - otherwise maybe the avatar will be unset if the user is changing things that aren't the avatar?

Comment (lucaslcode) [reply app/frontend/src/features/profile/ProfileForm.tsx]: But that might mess with the thumbnail...

**Review by lucaslcode — COMMENTED** [review]
Looks good, let's wait for backend

File: app/frontend/src/features/profile/ProfileForm.tsx (hunk 17 above)
Comment (lucaslcode) [reply app/frontend/src/features/profile/ProfileForm.tsx]: Maybe the solution would be an `initialImage` prop or something

**Comment by aapeliv** [comment]
>  The updateUser request gives an error when it includes the avatarKey

Fixed the typo

>  And some of the uploaded images gives "Internal Server Error" when I request them after upload

Fixed that one too, in #709.

Sorry, the backend has been incredibly buggy...

File: app/frontend/src/components/AvatarInput.tsx (hunk 16 above)
Comment (MichaelGuldborg) [reply app/frontend/src/components/AvatarInput.tsx]: That's pretty cool actually 🤔

File: app/frontend/src/features/profile/ProfileForm.tsx (hunk 17 above)
Comment (MichaelGuldborg) [reply app/frontend/src/features/profile/ProfileForm.tsx]: That's a good point.
It seems that the backend needs the avatarkey to update the profile image. I assumed that if it was undefined or empty then the backend would just ignore it.

@aapeliv

File: app/frontend/src/components/AvatarInput.tsx (hunk 15 above)
Comment (MichaelGuldborg) [reply app/frontend/src/components/AvatarInput.tsx]: I did have some issues when uploading the same image twice, which is why I added it.
But I'll check if it can be omitted

File: app/frontend/src/service/user.ts [hunk 18]
```diff
@@ -129,6 +130,7 @@ export async function updateProfile(
 ): Promise<Empty> {
   const req = new UpdateProfileReq();
 
+  const avatarKey = new wrappers.StringValue().setValue(profile.avatarKey);
```
Comment (lucaslcode) [inline app/frontend/src/service/user.ts]:
```suggestion
  const avatarKey = profile.avatarKey ? new wrappers.StringValue().setValue(profile.avatarKey) : undefined;
```

This might solve this - the purpose of the wrapper is to differentiate between unsetting (ie. setting to nothing - wrapper is present but value is null) and just not making any changes (wrapper itself not present). This should do since you can't clear the image at this point.

**Comment by lucaslcode** [comment]
Currently upload works, but changing the profile without changing the image causes an unknown backend error. I suspect there is no check for avatarKey existing but having a blank value @aapeliv .

File: app/frontend/src/service/user.ts (hunk 18 above)
Comment (MichaelGuldborg) [reply app/frontend/src/service/user.ts]: I've changed it to your suggestion

**Comment by lucaslcode** [comment]
@MichaelGuldborg  It's disappeared from the profile edit page - did it get lost in the merge?

File: app/frontend/src/components/ImageInput.tsx [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        name={name}
+        type="file"
+        onChange={async (event) => {
+          if (!event.target.files?.length) {
+            console.log("No files selected");
+            return;
+          }
+          const file = event.target.files[0];
+          const response = await mutation.mutateAsync(file);
+          onChange && onChange(response);
```
Comment (darrenvong) [inline app/frontend/src/components/ImageInput.tsx]:
```suggestion
          onChange(response);
```

File: app/frontend/src/components/ImageInput.tsx [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+export interface ImageInputProps {
+  id?: string;
+  name?: string;
+  onChange: (values: ImageInputValues) => void;
+  children: React.ReactNode;
+}
+
+export function ImageInput({ id, name, children, onChange }: ImageInputProps) {
+  const classes = useStyles();
+  const inputId = id + "-file-input";
```
Comment (darrenvong) [inline app/frontend/src/components/ImageInput.tsx]:
```suggestion
  const inputId = `${id}-file-input`;
```

File: app/frontend/src/components/ImageInput.tsx [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+    <>
+      <input
+        className={classes.input}
+        accept="image/*"
+        id={inputId}
+        name={name}
+        type="file"
+        onChange={async (event) => {
+          if (!event.target.files?.length) {
+            console.log("No files selected");
```
Comment (darrenvong) [inline app/frontend/src/components/ImageInput.tsx]:
```suggestion
```

File: app/frontend/src/features/profile/edit/EditProfile.tsx [hunk 22]
```diff
@@ -122,6 +127,23 @@ export default function EditProfileForm() {
       {user ? (
         <>
           <form onSubmit={onSubmit}>
+            <Controller
+              name={"avatarKey"}
```
Comment (darrenvong) [inline app/frontend/src/features/profile/edit/EditProfile.tsx]:
```suggestion
              name="avatarKey"
```

File: app/frontend/src/features/profile/edit/EditProfile.tsx [hunk 23]
```diff
@@ -122,6 +127,23 @@ export default function EditProfileForm() {
...
+              control={control}
+              render={({ onChange }) => {
+                return (
+                  <AvatarInput
+                    className={classes.avatar}
+                    id={"profile-picture"}
+                    name={"avatarKey"}
```
Comment (darrenvong) [inline app/frontend/src/features/profile/edit/EditProfile.tsx]:
```suggestion
                    id="profile-picture"
                    name="avatarKey"
```

File: app/frontend/src/components/AvatarInput.tsx (hunk 16 above)
Comment (darrenvong) [reply app/frontend/src/components/AvatarInput.tsx]: This is unnecessary since `onChange` is guaranteed by the interface to always exist?

File: app/frontend/src/components/AvatarInput.tsx (hunk 15 above)
Comment (darrenvong) [reply app/frontend/src/components/AvatarInput.tsx]: So is it safe to actually omit this? Seems a bit hacky to me if this can be resolved by the backend instead.

File: app/frontend/src/service/api.ts [hunk 24]
```diff
@@ -48,3 +51,23 @@ export async function ping() {
 
   return response.toObject();
 }
+
+export async function uploadFile(file: File): Promise<ImageInputValues> {
```
Comment (darrenvong) [inline app/frontend/src/service/api.ts:62]: Not sure if this manual work is necessary 🤔  We can always get the URL for this in a `useEffect` and populate the form's action that way?

I feel like it'd be simpler to use HTML form to handle this as it already deals with form data properly out of the box.

File: app/frontend/src/service/api.ts [hunk 25]
```diff
@@ -1,11 +1,14 @@
...
+import { ImageInputValues } from "../components/ImageInput";
 import {
   CancelFriendRequestReq,
   PingReq,
   RespondFriendRequestReq,
   SendFriendRequestReq,
 } from "pb/api_pb";
-import client from "service/client";
+
+import client from "./client";
```
Comment (darrenvong) [inline app/frontend/src/service/api.ts]: Build is failing on the import order of these.

**Review by darrenvong — CHANGES_REQUESTED** [review]
Also out of curiosity once this is merged, how do you use this locally?

**Comment by darrenvong** [comment]
Also this is crashing with `Error: There is already a source with this ID`, which I suspect your Map fix will fix this?

File: app/frontend/src/service/api.ts (hunk 24 above)
Comment (lucaslcode) [reply app/frontend/src/service/api.ts:62]: I think it's better this way, since otherwise you would be making a media upload request every time you go to edit anything on your profile, plus an html form `submit()` doesn't return a promise, so it will end up more complicated.

**Comment by lucaslcode** [comment]
image

To test locally you need to run media server docker container, but it's only in the production docker-compose file atm

File: app/frontend/src/components/AvatarInput.tsx (hunk 16 above)
Comment (darrenvong) [reply app/frontend/src/components/AvatarInput.tsx]:
```suggestion
    onChange(values.key);
```

File: app/frontend/src/service/api.ts (hunk 24 above)
Comment (darrenvong) [reply app/frontend/src/service/api.ts:62]: Does this need to be part of the main form though if this is a standalone thing? It shouldn't be sent every time if it's in its own form. What would you need the promise for in that case?

File: app/frontend/src/components/ImageInput.tsx [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+        className={classes.input}
+        accept="image/*"
+        id={inputId}
+        name={name}
+        type="file"
+        onChange={async (event) => {
+          if (!event.target.files?.length) return;
+          const file = event.target.files[0];
+          const response = await mutation.mutateAsync(file);
+          onChange(response);
```
Comment (darrenvong) [inline app/frontend/src/components/ImageInput.tsx]: Not sure auto submitting like this is good UX - what if the user selects the wrong image and want to correct it?

Comment (lucaslcode) [reply app/frontend/src/components/ImageInput.tsx]: They can choose a new image - it does auto-upload, but the upload will only be placed in the resource (in the case, profile) when the form as a whole is submitted.

Comment (lucaslcode) [reply app/frontend/src/components/ImageInput.tsx]: What I mean is, this auto-submits and auto-receives a key, but the key isn't put into the profile automatically, just into the form state.

Comment (lucaslcode) [reply app/frontend/src/components/ImageInput.tsx]: If you're content with that and you approve, feel free to merge this pr

Comment (darrenvong) [reply app/frontend/src/components/ImageInput.tsx]: Ahh I see, sounds like more of a problem for Aapeli's wallet on S3 bucket cost than actual technical problem here for me here 😂

Comment (darrenvong) [reply app/frontend/src/components/ImageInput.tsx]: Although storage (half) joke aside, it's still a bit shitty for people on mobile data for this to always upload?

**Comment by lucaslcode** [comment]
image

**Comment by lucaslcode** [comment]
How's this, will do some tests if it's ok

File: app/frontend/src/components/AvatarInput.tsx [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+      reader.readAsDataURL(file);
+    });
+    setImageUrl(base64);
+    setFile(file);
+  };
+
+  const handleConfirm = async () => {
+    const response = await mutation.mutateAsync();
+    //const randomInt = Math.floor(Math.random() * 100); // force reload onChange
+    //setImageUrl(response.thumbnail_url + "?rand=" + randomInt);
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.tsx]:
```suggestion
```

File: app/frontend/src/components/AvatarInput.tsx [hunk 28]
```diff
@@ -1,54 +1,166 @@
...
+      validate: () => !isConfirming || UPLOAD_PENDING_ERROR,
+    },
+  });
+
+  const handleChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (!event.target.files?.length) return;
+    const file = event.target.files[0];
+    const base64 = await new Promise<string>((resolve, reject) => {
+      const reader = new FileReader();
+      reader.onload = () => resolve(reader!.result as string);
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.tsx]:
```suggestion
      reader.onload = () => resolve(reader.result as string);
```

File: app/frontend/src/components/constants.ts [hunk 29]
```diff
@@ -0,0 +1,7 @@
+export const CANCEL_UPLOAD = "Cancel upload";
+export const CONFIRM_UPLOAD = "Confirm upload";
+export const getAvatarLabel = (name: string) => `${name} avatar`;
```
Comment (darrenvong) [inline app/frontend/src/components/constants.ts]:
```suggestion
export const getAvatarLabel = (name: string) => `${name}'s profile picture`;
```

File: app/frontend/src/components/AvatarInput.stories.tsx [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+import { useForm } from "react-hook-form";
+
+import AvatarInput from "./AvatarInput";
+import Button from "./Button";
+
+export default {
+  component: AvatarInput,
+  title: "Components/Compound/AvatarInput",
+} as Meta;
+
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.stories.tsx:13]: This is broken when pressing confirm. I think `uploadFile` from service in this will need mocking.

**Review by darrenvong — CHANGES_REQUESTED** [review]
Pretty minor things left in the comment now, but this looks fine to me. Go ahead and add the tests if you are planning to 👍

File: app/frontend/src/components/AvatarInput.tsx (hunk 28 above)
Comment (lucaslcode) [reply app/frontend/src/components/AvatarInput.tsx]: Typescript doesn't know it can't be undefined, since it starts as undefined and the event can be triggered from anything in theory

Comment (darrenvong) [reply app/frontend/src/components/AvatarInput.tsx]: `reader` won't be undefined though as you instantiated on the line above? I checked locally and it didn't give me an error without it.

Comment (lucaslcode) [reply app/frontend/src/components/AvatarInput.tsx]: Oh quite right

File: app/frontend/src/components/AvatarInput.stories.tsx (hunk 30 above)
Comment (lucaslcode) [reply app/frontend/src/components/AvatarInput.stories.tsx:13]: I used `ChangeEmail.stories.tsx` as a mock example but it isn't mocking... any idea why?

Comment (darrenvong) [reply app/frontend/src/components/AvatarInput.stories.tsx:13]: Maybe it's because `AvatarInput` is importing `service` instead of `service/index`?

File: app/frontend/src/components/AvatarInput.test.tsx [hunk 31]
```diff
@@ -0,0 +1,175 @@
...
+      MOCK_INITIAL_SRC
+    );
+
+    userEvent.upload(screen.getByLabelText(SELECT_AN_IMAGE), MOCK_FILE);
+
+    await waitFor(() => {
+      expect(screen.getByLabelText(CONFIRM_UPLOAD)).toBeVisible();
+    });
+
+    userEvent.click(screen.getByLabelText(CONFIRM_UPLOAD));
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.test.tsx]: Is this not meant to click `CANCEL_UPLOAD` instead based on the test description?

File: app/frontend/src/components/AvatarInput.test.tsx [hunk 32]
```diff
@@ -0,0 +1,175 @@
...
+    await waitFor(() => {
+      expect(uploadFileMock).toHaveBeenCalled();
+    });
+
+    userEvent.click(screen.getByRole("button", { name: SUBMIT }));
+
+    await waitFor(() => {
+      expect(submitForm).toHaveBeenCalledWith({ avatarInput: MOCK_KEY });
+    });
+  });
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.test.tsx:85]: Check the image src here has also been updated to the new src?

File: app/frontend/src/components/AvatarInput.test.tsx [hunk 33]
```diff
@@ -0,0 +1,175 @@
...
+
+    await waitFor(() => {
+      expect(uploadFileMock).toHaveBeenCalledTimes(2);
+    });
+
+    expect(screen.getByAltText(getAvatarLabel(NAME))).not.toHaveProperty(
+      "src",
+      MOCK_INITIAL_SRC
+    );
+
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.test.tsx]:
```suggestion
    expect(screen.getByAltText(getAvatarLabel(NAME))).toHaveProperty(
      "src",
      "thumb0.jpg"
    );
```
Checking it has the src of the first upload implies it's not `MOCK_INITIAL_SRC`

File: app/frontend/src/components/AvatarInput.test.tsx [hunk 34]
```diff
@@ -0,0 +1,175 @@
...
+      expect(submitForm).toHaveBeenCalledWith({ avatarInput: MOCK_KEY });
+    });
+  });
+
+  it("doesn't submit without confirming/cancelling", async () => {
+    userEvent.upload(screen.getByLabelText(SELECT_AN_IMAGE), MOCK_FILE);
+
+    await waitFor(() => {
+      expect(screen.getByLabelText(CONFIRM_UPLOAD)).toBeVisible();
+    });
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.test.tsx]:
```suggestion
      expect(await screen.findByLabelText(CONFIRM_UPLOAD)).toBeVisible();
```
For single assertion like this, you can use `findBy*` instead. It's shorthand equivalent to what you got there.

File: app/frontend/src/components/AvatarInput.test.tsx [hunk 35]
```diff
@@ -0,0 +1,175 @@
...
+
+    userEvent.click(screen.getByRole("button", { name: SUBMIT }));
+
+    await waitFor(() => {
+      expect(screen.getByText(UPLOAD_PENDING_ERROR)).toBeVisible();
+    });
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.test.tsx]: Same comment as above

File: app/frontend/src/components/AvatarInput.test.tsx [hunk 36]
```diff
@@ -0,0 +1,175 @@
...
+
+    await waitFor(() => {
+      expect(screen.getByLabelText(CONFIRM_UPLOAD)).toBeVisible();
+      expect(screen.getByLabelText(CANCEL_UPLOAD)).toBeVisible();
+    });
+
+    userEvent.click(screen.getByLabelText(CONFIRM_UPLOAD));
+
+    await waitFor(() => {
+      expect(uploadFileMock).toHaveBeenCalled();
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.test.tsx]:
```suggestion
      expect(uploadFileMock).toHaveBeenCalledTimes(1);
```
So we know it's not going rogue and being called unnecessarily somehow?

File: app/frontend/src/components/AvatarInput.test.tsx [hunk 37]
```diff
@@ -0,0 +1,175 @@
...
+    );
+  });
+
+  it("uploads upon confirmation and submits key", async () => {
+    userEvent.upload(screen.getByLabelText(SELECT_AN_IMAGE), MOCK_FILE);
+
+    await waitFor(() => {
+      expect(screen.getByLabelText(CONFIRM_UPLOAD)).toBeVisible();
+      expect(screen.getByLabelText(CANCEL_UPLOAD)).toBeVisible();
+    });
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.test.tsx]:
```suggestion
    expect(await screen.findByLabelText(CONFIRM_UPLOAD)).toBeVisible();
```
Since both of these buttons render together, it's ok to only wait for one of them, since that'd be long enough for the other to appear too.

**Review by darrenvong — CHANGES_REQUESTED** [review]
Some solid testing effort there 💪

**Comment by darrenvong** [comment]
Also tests for negative cases? i.e. what happens if an invalid file is given, and when the file reader throws an error

**Comment by lucaslcode** [comment]
@darrenvong as far as I can tell, there's no way to check for valid images. However the server does and I added error handling.

File: app/frontend/src/components/AvatarInput.test.tsx [hunk 38]
```diff
@@ -0,0 +1,165 @@
...
+    expect(submitForm).not.toHaveBeenCalled();
+  });
+
+  it.skip("displays an error for an invalid file", async () => {
+    userEvent.upload(
+      screen.getByLabelText(SELECT_AN_IMAGE),
+      new File([new Blob(undefined)], "")
+    );
+    expect(await screen.findByText(INVALID_FILE)).toBeVisible();
+  });
```
Comment (lucaslcode) [inline app/frontend/src/components/AvatarInput.test.tsx]: No matter what I tried, couldn't make an invalid file to test this....

Comment (darrenvong) [reply app/frontend/src/components/AvatarInput.test.tsx]: I got it working by making `readAsDataURL` emit an `error` event on all `FileReader` instances for this test specifically, which I guess is what happens when there's an error as these seem to use the event callback pattern rather than throwing an error directly, so I think it's fine to do? More interested in hitting the error handling path rather than how it's achieved here...

```suggestion
  it("displays an error for an invalid file", async () => {
    jest.spyOn(FileReader.prototype, "readAsDataURL").mockImplementation(() => {
      FileReader.prototype.dispatchEvent(new Event("error"));
    });
    userEvent.upload(
      screen.getByLabelText(SELECT_AN_IMAGE),
      new File([new Blob(undefined)], "")
    );
    expect(await screen.findByText(INVALID_FILE)).toBeVisible();
  });
```

**Comment by darrenvong** [comment]
If the server will throw an error, couldn't we simulate this since those are already mocked? What happens if `handleConfirm` throws from the `mutateAsync` call?

File: app/frontend/src/components/AvatarInput.tsx [hunk 39]
```diff
@@ -0,0 +1,173 @@
...
+      });
+      setImageUrl(base64);
+      setFile(file);
+    } catch {
+      setReaderError(INVALID_FILE);
+    }
+  };
+
+  const handleConfirm = async () => {
+    const response = await mutation.mutateAsync();
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.tsx]: What happens if this throws? I think the whole UI will probably crash here...

Comment (lucaslcode) [reply app/frontend/src/components/AvatarInput.tsx]: Interestingly not, but I would have expected it to! Wrapped it in try just in case.

File: app/frontend/src/components/AvatarInput.tsx [hunk 40]
```diff
@@ -105,11 +105,15 @@ export function AvatarInput({
...
-    confirmedUpload.current = response;
-    setFile(null);
+    try {
+      const response = await mutation.mutateAsync();
+      field.onChange(response.key);
+      setImageUrl(response.thumbnail_url);
+      confirmedUpload.current = response;
+      setFile(null);
+    } catch {
+      //no need to do anything here, error goes to mutation.error
```
Comment (darrenvong) [inline app/frontend/src/components/AvatarInput.tsx]: Would it be easier to use `mutate` (the non async awaitable version) in that case? 🤔 
Then we wouldn't need this catch at all as it'd still be handled by React Query's mutation error?

**Review by darrenvong — APPROVED** [review]
